mat/io.go

```diff
@@ -198,23 +239,32 @@ func (m *Dense) UnmarshalBinaryFrom(r io.Reader) (int, error) {
 		m.mat.Data[i] = math.Float64frombits(binary.LittleEndian.Uint64(buf[:]))
 	}
 
+	if n < int(size)*sizeFloat64+2*sizeInt64+2 {
+		return n, io.ErrUnexpectedEOF
+	}
+
 	return n, nil
 }
 
 // MarshalBinary encodes the receiver into a binary form and returns the result.
 //
 // VecDense is little-endian encoded as follows:
-//   0 -  7  number of elements     (int64)
-//   8 - ..  vector's data elements (float64)
+//
+//   0       'D' - type annotation  (int8)
```

Comment:
Does this extend to including vectors in that?

The scheme that looks like needs to exist is:

|Type|Form|Packing|Uplo|Unit|Rows|Columns|kU |kL |
|---|---|---|---|---|---|---|---|---|
| |`uint8` [GST]|`uint8` [BPF]|`uint8` [AUL]|`bool`|`int64`|`int64`|`int64`|`int64`|
|General|'G'|'F'|'A'|`false`|`r`|`c`|0|0|
|Band|'G'|'B'|'A'|`false`|`r`|`c`|`kU`|`kL`|
|Symmetric|'S'|'F'|`ul`|`false`|`n`|`n`|0|0|
|SymmetricBand|'S'|'B'|`ul`|`false`|`r`|`c`|`k`|`k`|
|SymmetricPacked|'S'|'P'|`ul`|`false`|`r`|`c`|0|0|
|Triangular|'T'|'F'|`ul`|`Diag==Unit`|`n`|`n`|0|0|
|TriangularBand|'T'|'B'|`ul`|`Diag==Unit`|`r`|`c`|`k`|`k`|
|TriangularPacked|'T'|'P'|`ul`|`Diag==Unit`|`r`|`c`|0|0|

This can be unmarshaled into a common internal struct
```
type storage struct {
    Form    byte
    Packing byte
    Uplo    byte
    Unit    bool
    Rows    int
    Cols    int
    KU      int
    KL      int
}
```

Does this sound reasonable?